src/ejabberd_config.erl

```diff
@@ -804,212 +311,196 @@ get_vh_by_auth_method(AuthMethod) ->
-        {ok, FileInfo} ->
-            case {FileInfo#file_info.type, FileInfo#file_info.access} of
-                {regular, read} -> true;
-                {regular, read_write} -> true;
-                _ -> false
-            end;
-        {error, _Reason} ->
-            false
-    end.
+handle_table_does_not_exist_error(Table) ->
+    MnesiaDirectory = mnesia:system_info(directory),
+    ?ERROR_MSG("Error reading Mnesia database spool files:~n"
+               "The Mnesia database couldn't read the spool file for the table '~p'.~n"
+               "ejabberd needs read and write access in the directory:~n   ~s~n"
+               "Maybe the problem is a change in the computer hostname, ~n"
+               "or a change in the Erlang node name, which is currently:~n   ~p~n"
+               "Check the ejabberd guide for details about changing the~n"
+               "computer hostname or Erlang node name.~n",
+               [Table, MnesiaDirectory, node()]),
+    exit("Error reading Mnesia database").
 
 %%--------------------------------------------------------------------
 %% Configuration reload
 %%--------------------------------------------------------------------
 -spec parse_file(file:name()) -> state().
 parse_file(ConfigFile) ->
     Terms = get_plain_terms_file(ConfigFile),
-    State = lists:foldl(fun search_hosts_and_pools/2, #state{}, Terms),
-    TermsWExpandedMacros = replace_macros(Terms),
-    lists:foldl(fun process_term/2,
-                add_option(odbc_pools, State#state.odbc_pools, State),
-                TermsWExpandedMacros).
+    mongoose_config:parse_terms(Terms).
 
 -spec reload_local() -> {ok, iolist()} | no_return().
 reload_local() ->
-    ConfigFile = get_ejabberd_config_path(),
-    State0 = parse_file(ConfigFile),
-    {CC, LC, LHC} = get_config_diff(State0),
-    ConfigVersion = compute_config_version(get_local_config(),
-                                           get_host_local_config()),
-    State1 = State0#state{override_global = true,
-                          override_local  = true,
-                          override_acls   = true},
-    try
-        {ok, _} = apply_changes(CC, LC, LHC, State1, ConfigVersion),
-        ?WARNING_MSG("node config reloaded from ~s", [ConfigFile]),
-        {ok, io_lib:format("# Reloaded: ~s", [node()])}
-    catch
-        Error:Reason ->
-            Msg = msg("failed to apply config on node: ~p~nreason: ~p",
-                      [node(), {Error, Reason}]),
-            ?WARNING_MSG("node config reload failed!~n"
-                         "current config version: ~p~n"
-                         "config file: ~s~n"
-                         "reason: ~p~n"
-                         "stacktrace: ~ts",
-                         [ConfigVersion, ConfigFile, Msg,
-                          msg("~p", [erlang:get_stacktrace()])]),
-            error(Msg)
+    NodeStates = [config_state()],
+    ReloadStrategy = mongoose_config:cluster_reload_strategy(NodeStates),
```

Comment:
I wanted it to be an atom, but it ended up a bit bigger. :trollface: 
Basically, strategy in MAM was introduced, so no more decisions are made in code, that receives strategy. Kinda similar here. 

Not really state, because you can't change it, once it's generated.

TLDR:

```erlang
Strategy = make_decisions(),
execute_decisions(Strategy)

or

Strategy = smart_pure_stuff(current_env_state()),
stupid_real_world_stuff(Strategy)
```

Overall, idea is not to make any decisions, once we started doing destructive changes.
So, you can call for strategy any time, but don't execute it (for `dryrun`, for validation, for debugging).